common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/NonceRedirectHandler.kt

```diff
@@ -0,0 +1,89 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package com.microsoft.identity.common.internal.ui.webview.challengehandlers
+
+import android.app.Activity
+import android.webkit.WebView
+import com.microsoft.identity.common.CommonRefreshTokenCredentialProvider
+import com.microsoft.identity.common.adal.internal.AuthenticationConstants
+import com.microsoft.identity.common.adal.internal.util.StringExtensions
+import com.microsoft.identity.common.logging.Logger
+import java.net.URL
+
+/**
+ * Handler for processing nonce from redirect and attaching new prt credential header on web view.
+ */
+class NonceRedirectHandler(
+    private val webview: WebView? = null,
+    private val headers: HashMap<String, String>
+) : IChallengeHandler<URL, Void> {
+    private val TAG = NonceRedirectHandler::class.java.simpleName
+
+    override fun processChallenge(input: URL) : Void? {
+        val nonce = getNonceFromRedirectUrl(input)
+        if (nonce != null) {
+            modifyHeadersWithNewRefreshTokenCredential(nonce, input.toString())
+        }
+        webview?.loadUrl(input.toString(), headers)
+        return null
+    }
+
+    private fun getNonceFromRedirectUrl(url: URL): String? {
+        val parameters = StringExtensions.getUrlParameters(url.toString())
+        return parameters["sso_nonce"]
+    }
+
+    private fun getPrtHeader(requestHeaders: HashMap<String, String>): String? {
+        return requestHeaders[AuthenticationConstants.Broker.PRT_RESPONSE_HEADER]
+    }
+
+    // Updates the headers by attaching a new refresh token credential header (Generated using the new nonce).
+    private fun modifyHeadersWithNewRefreshTokenCredential(
+        nonce: String,
+        url: String
+    ) {
+        val methodTag = "$TAG:getHeadersWithNewRefreshTokenCredential"
+        val prtHeader = getPrtHeader(headers)
+        if (!prtHeader.isNullOrEmpty()) {
+            Logger.info(methodTag, "PRT credential header found in headers!")
+            val username = getUserNameFromWebViewUrl(url)
+            if (username != null) {
+                val updatedRefreshTokenCredentialHeader =
+                    CommonRefreshTokenCredentialProvider.getRefreshTokenCredentialUsingNewNonce(
+                        url, username,
```

Comment:
>url [](http://example.com/codeflow?start=24&length=3)

pass authority url itself instead of input url

Reply:
Authority URL needs to be extracted out of input URL. Is there any reason why we need this?

Reply:
Addressed this by changing the parameter name to inputUrl to avoid any confusion.